pylint/checkers/typecheck.py

```diff
@@ -1027,11 +1028,26 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
                 return
 
             try:
-                if not [
-                    n
-                    for n in owner.getattr(node.attrname)
-                    if not isinstance(n.statement(future=True), nodes.AugAssign)
-                ]:
+                for n in owner.getattr(node.attrname):
+                    n_parent = n.parent
+                    found_attribute = (
+                        # Skip augmented assignments
+                        not isinstance(n.statement(future=True), nodes.AugAssign)
+                        # Skip self-referencing assignments
+                        and n_parent is not node.parent
+                        # Skip assignments that are only type annotations
+                        and not (
+                            isinstance(n_parent, nodes.AnnAssign)
+                            and n_parent.value is None
+                        )
+                        # But allow dataclasses
+                        or brain_dataclasses.is_decorated_with_dataclass(
+                            n_parent.parent
+                        )
+                    )
+                    if found_attribute:
+                        break
```

Comment:
```suggestion
                    # Skip augmented assignments
                    if isinstance(n.statement(future=True), nodes.AugAssign):
                        break
                    # Skip self-referencing assignments
                    if not n_parent is not node.parent:
                        break
                    # Skip assignments that are only type annotations
                    if isinstance(n_parent, nodes.AnnAssign) and n_parent.value is None:
                        break
                    # Allow dataclasses
                    if brain_dataclasses.is_decorated_with_dataclass(n_parent.parent):
                        break
```

Would this work? Refactoring for loops on Github is hard so I might have misplaced a not or if, but I think this should work right? Personally I think using premature break is easier on the eyes, but opinions might differ obviously.

Reply:
Ah I don't think so, because `found_attribute` is three conditions, and they all need to pass, then we have to mix `continue` and `break`, like this:

```python
            try:
                for n in owner.getattr(node.attrname):
                    n_parent = n.parent
                    # Skip augmented assignments
                    if isinstance(n.statement(future=True), nodes.AugAssign):
                        continue
                    # Skip self-referencing assignments
                    if n_parent is node.parent:
                        continue
                    # Skip assignments that are only type annotations and not dataclasses
                    if (
                        isinstance(n_parent, nodes.AnnAssign)
                        and n_parent.value is None
                        and not brain_dataclasses.is_decorated_with_dataclass(
                            n_parent.parent
                        )
                    ):
                        continue
                    break
                else:
                    missingattr.add((owner, name))
                    continue
```

I'm 50/50 on whether that's any better, since we end up with different nesting levels of `continue`s.


***

However, thank you for having me look at this again, I've noticed missing parentheses at the end to make what's now 4 conditions into what was meant to be the 3 conditions.

Reply:
Personally I think this is better and we can even improve on it some more!

For example, the `StatementMissing` catcher can be move to the `statement()` line. Perhaps that's a bit beyond the scope of this PR but I think using this broken down version allows for more refactoring.

Reply:
I'll push with what I quoted above, and let's look into moving the excepts closer in a follow-up PR, because it's possible there could be a behavior change -- we could simply continue searching nodes when handling an exception rather than bail out entirely, as before.